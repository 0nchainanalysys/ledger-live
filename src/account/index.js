// @flow
import type {
  AddAccountsSection,
  AddAccountsSectionResult
} from "./addAccounts";
import {
  getMainAccount,
  getAccountCurrency,
  getAccountUnit,
  isAccountEmpty,
  clearAccount,
  flattenAccounts,
  addPendingOperation,
  shortAddressPreview
} from "./helpers";
import {
  shouldShowNewAccount,
  canBeMigrated,
  findAccountMigration,
  groupAddAccounts,
  addAccounts,
  migrateAccounts
} from "./addAccounts";
import {
  inferSubOperations,
  toOperationRaw,
  fromOperationRaw,
  toTokenAccountRaw,
  fromTokenAccountRaw,
  toAccountRaw,
  fromAccountRaw
} from "./serialization";
import { encodeAccountId, decodeAccountId, getWalletName } from "./accountId";
import {
  getAccountPlaceholderName,
  getNewAccountPlaceholderName,
  validateNameEdition
} from "./accountName";
import {
  sortAccountsComparatorFromOrder,
  comparatorSortAccounts,
  flattenSortAccounts,
  nestedSortAccounts
} from "./ordering";
import {
  groupAccountsOperationsByDay,
  groupAccountOperationsByDay
} from "./groupOperations";

export type { AddAccountsSection, AddAccountsSectionResult };

export {
  getMainAccount,
  getAccountCurrency,
  getAccountUnit,
  isAccountEmpty,
  clearAccount,
  flattenAccounts,
  shouldShowNewAccount,
  canBeMigrated,
  findAccountMigration,
  groupAddAccounts,
  addAccounts,
  migrateAccounts,
  inferSubOperations,
  toOperationRaw,
  fromOperationRaw,
  toTokenAccountRaw,
  fromTokenAccountRaw,
  toAccountRaw,
  fromAccountRaw,
  encodeAccountId,
  decodeAccountId,
  getWalletName,
  getAccountPlaceholderName,
  getNewAccountPlaceholderName,
  validateNameEdition,
  sortAccountsComparatorFromOrder,
  comparatorSortAccounts,
  flattenSortAccounts,
  nestedSortAccounts,
  groupAccountsOperationsByDay,
  groupAccountOperationsByDay,
  addPendingOperation,
  shortAddressPreview
};
