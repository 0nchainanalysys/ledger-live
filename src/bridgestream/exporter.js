// @flow
import type { Account } from "../types/account";

export type DataIn = {
  // accounts to export (filter them to only be the visible ones)
  accounts: Account[],
  // the name of the exporter. e.g. "desktop" for the desktop app
  exporterName: string,
  // the version of the exporter. e.g. the desktop app version
  exporterVersion: string,
  // enable a mode that will add empty space to make all chunks of same size. to makes QR code of same dimension
  pad?: boolean
};

/**
 * export data into a chunk of string
 * @memberof bridgestream/exporter
 */
export function makeChunks({
  accounts,
  exporterName,
  exporterVersion,
  pad
}: DataIn): string[] {
  const chunksFormatVersion = 2;
  const data = [
    ["meta", chunksFormatVersion, exporterName, exporterVersion],
    ...accounts.map(account => [
      "account",
      account.id,
      account.name,
      account.currency.id
    ])
  ];
  let r = data.map((arr, i) => JSON.stringify([data.length, i, ...arr]));
  if (pad) {
    const max = r.reduce((max, s) => Math.max(max, s.length), 0);
    r = r.map(s => s.padEnd(max));
  }
  return r;
}
