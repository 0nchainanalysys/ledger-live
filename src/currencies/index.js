// @flow

import { encodeURIScheme, decodeURIScheme } from "./CurrencyURIScheme";
import type { Currency } from "../types";

import { sanitizeValueString } from "./sanitizeValueString";

import { sortByMarketcap } from "./sortByMarketcap";

import {
  listFiatCurrencies,
  findFiatCurrencyByTicker,
  getFiatCurrencyByTicker,
  hasFiatCurrencyTicker
} from "../data/fiat";

import {
  listCryptoCurrencies,
  getCryptoCurrencyById,
  hasCryptoCurrencyId,
  findCryptoCurrency,
  findCryptoCurrencyById,
  findCryptoCurrencyByScheme,
  findCryptoCurrencyByTicker
} from "../data/cryptocurrencies";

import {
  listTokens,
  listTokensForCryptoCurrency,
  findTokenByTicker,
  findTokenById,
  findTokenByAddress,
  hasTokenId,
  getTokenById
} from "../data/tokens";

import { parseCurrencyUnit } from "./parseCurrencyUnit";

import { chopCurrencyUnitDecimals } from "./chopCurrencyUnitDecimals";

import {
  formatCurrencyUnit,
  formatCurrencyUnitFragment
} from "./formatCurrencyUnit";

import { formatShort } from "./formatShort";

import { valueFromUnit } from "./valueFromUnit";

import { getCurrencyColor } from "./color";

const findCurrencyByTicker = (ticker: string): ?Currency =>
  findCryptoCurrencyByTicker(ticker) ||
  findFiatCurrencyByTicker(ticker) ||
  findTokenByTicker(ticker);

export {
  sortByMarketcap,
  listFiatCurrencies,
  listCryptoCurrencies,
  getFiatCurrencyByTicker,
  findCurrencyByTicker,
  findCryptoCurrency,
  findCryptoCurrencyById,
  findCryptoCurrencyByTicker,
  findCryptoCurrencyByScheme,
  findFiatCurrencyByTicker,
  hasFiatCurrencyTicker,
  listTokens,
  listTokensForCryptoCurrency,
  findTokenByAddress,
  findTokenByTicker,
  findTokenById,
  hasTokenId,
  getTokenById,
  parseCurrencyUnit,
  chopCurrencyUnitDecimals,
  formatCurrencyUnit,
  formatCurrencyUnitFragment,
  formatShort,
  getCryptoCurrencyById,
  hasCryptoCurrencyId,
  encodeURIScheme,
  decodeURIScheme,
  valueFromUnit,
  sanitizeValueString,
  getCurrencyColor
};
