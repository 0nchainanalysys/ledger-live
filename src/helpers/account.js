// @flow
import type { Account, Unit } from "../types";
import type { CalculateCounterValue } from "./countervalue";

/*
*/
export type BalanceHistory = Array<{ date: Date, value: number }>;

/**
 * generate an array of {daysCount} datapoints, one per day,
 * for the balance history of an account.
 * The last item of the array is the balance available right now.
 */
export function getBalanceHistory(
  account: Account,
  daysCount: number
): BalanceHistory {
  const history = [];
  let { balance } = account;
  let i = 0; // index of operation
  let t = new Date();
  history.unshift({ date: t, value: balance });
  t = new Date(t.getFullYear(), t.getMonth(), t.getDate()); // start of the day
  for (let d = 0; d < daysCount; d++) {
    // accumulate operations after time t
    while (i < account.operations.length && account.operations[i].date > t) {
      balance -= account.operations[i].amount;
      i++;
    }
    history.unshift({ date: t, value: balance });
    t = new Date(t - 24 * 60 * 60 * 1000);
  }
  return history;
}

/**
 * calculate the total balance history for all accounts in a reference fiat unit
 * and using a CalculateCounterValue function (see countervalue helper)
 * NB the last item of the array is actually the current total balance.
 */
export function getBalanceHistorySum(
  accounts: Account[],
  daysCount: number,
  fiatUnit: Unit,
  calculateCounterValue: CalculateCounterValue
): BalanceHistory {
  if (accounts.length === 0) {
    const now = Date.now();
    const zeros: 0[] = Array(daysCount).fill(0);
    return zeros.map((value, i) => ({
      date: new Date(now - 24 * 60 * 60 * 1000 * (daysCount - i - 1)),
      value
    }));
  }
  return accounts
    .map(account => {
      const history = getBalanceHistory(account, daysCount);
      const calc = calculateCounterValue(account.currency, fiatUnit);
      return history.map(h => ({
        date: h.date,
        value: calc(h.value, h.date)
      }));
    })
    .reduce((acc, history) =>
      acc.map((a, i) => ({
        date: a.date,
        value: a.value + history[i].value
      }))
    );
}
