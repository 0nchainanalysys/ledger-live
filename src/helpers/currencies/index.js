//@flow

import { encodeURIScheme, decodeURIScheme } from "./CurrencyURIScheme";

import {
  listFiatCurrencies,
  getFiatCurrencyByTicker,
  hasFiatCurrencyTicker
} from "../../data/fiat";

import {
  listCryptoCurrencies,
  getCryptoCurrencyById,
  hasCryptoCurrencyId
} from "../../data/cryptocurrencies";

import { parseCurrencyUnit } from "./parseCurrencyUnit";

import { chopCurrencyUnitDecimals } from "./chopCurrencyUnitDecimals";

import {
  formatCurrencyUnit,
  formatCurrencyUnitFragment
} from "./formatCurrencyUnit";

import { formatShort } from "./formatShort";

export {
  listFiatCurrencies,
  listCryptoCurrencies,
  getFiatCurrencyByTicker,
  hasFiatCurrencyTicker,
  parseCurrencyUnit,
  chopCurrencyUnitDecimals,
  formatCurrencyUnit,
  formatCurrencyUnitFragment,
  formatShort,
  getCryptoCurrencyById,
  hasCryptoCurrencyId,
  encodeURIScheme,
  decodeURIScheme
};
