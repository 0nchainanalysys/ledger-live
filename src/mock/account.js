/**
 * @module mock/account
 * @flow
 */
import { listCurrencies } from "@ledgerhq/currencies";
import Prando from "prando";
import type { Currency } from "@ledgerhq/currencies";
import type { Account, Operation } from "../types";

const currencies = listCurrencies();

/**
 * @memberof mock/account
 */
export function genBitcoinAddressLike(rng: Prando) {
  const charset = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
  return `1${rng.nextString(rng.nextInt(25, 34), charset)}`;
}

/**
 * @memberof mock/account
 */
export function genHex(length: number, rng: Prando) {
  return rng.nextString(length, "0123456789ABCDEF");
}

/**
 * @memberof mock/account
 */
export function genAddress(currency: Currency, rng: Prando) {
  if (currency.coinType === 60 || currency.coinType === 61) {
    return `0x${genHex(40, rng)}`;
  }
  return genBitcoinAddressLike(rng);
}

// TODO fix the mock to never generate negative balance...
/**
 * @memberof mock/account
 */
export function genOperation(
  account: Account,
  ops: *,
  currency: Currency,
  rng: Prando
): Operation {
  const lastOp = ops[ops.length - 1];
  const date = new Date(
    (lastOp ? lastOp.date : Date.now()) -
      rng.nextInt(0, 100000000 * rng.next() * rng.next())
  );
  const address = genAddress(currency, rng);
  return {
    id: String(`mock_op_${ops.length}_${account.id}`),
    confirmations: 0,
    accountId: account.id,
    address,
    addresses: [address],
    amount:
      (rng.next() < 0.3 ? -1 : 1) *
      rng.nextInt(0, 100000000 * rng.next() * rng.next()),
    hash: genHex(64, rng),
    date,
    blockHeight:
      account.lastBlockHeight - Math.floor((Date.now() - date) / 900000)
  };
}

/**
 * @memberof mock/account
 */
export function genAddingOperationsInAccount(
  account: Account,
  count: number,
  seed: number | string
): Account {
  const rng = new Prando(seed);
  const copy: Account = { ...account };
  copy.operations = Array(count)
    .fill(null)
    .reduce(ops => {
      const op = genOperation(copy, ops, copy.currency, rng);
      return ops.concat(op);
    }, copy.operations);
  ensureNoNegative(copy.operations);
  copy.balance = account.operations.reduce((sum, op) => sum + op.amount, 0);
  return copy;
}

/**
 * @param id is a number or a string, used as an account identifier and as a seed for the generation.
 * @memberof mock/account
 */
type GenAccountOptions = {
  operationsSize?: number
};

export function genAccount(
  id: number | string,
  opts: GenAccountOptions = {}
): Account {
  const rng = new Prando(id);
  const currency = rng.nextArrayItem(currencies);
  const operationsSize = opts.operationsSize || rng.nextInt(1, 200);
  const address = genAddress(currency, rng);
  const account = {
    id: `mock_account_${id}`,
    coinType: currency.coinType,
    index: 0,
    path: "49'/1'/1'/0/2",
    rootPath: "49'/1'/1'",
    xpub: genHex(64, rng),
    archived: false,
    currency,
    lastBlockHeight: rng.nextInt(100000, 200000),
    unit: rng.nextArrayItem(currency.units),
    balance: 0,
    address,
    addresses: [address],
    operations: [],
    operationsSize,
    minConfirmations: 2,
    name: rng.nextString(rng.nextInt(4, 34)),
    balanceByDay: {}
  };

  account.operations = Array(operationsSize)
    .fill(null)
    .reduce(ops => {
      const op = genOperation(account, ops, currency, rng);
      return ops.concat(op);
    }, []);

  ensureNoNegative(account.operations);

  account.balance = account.operations.reduce((sum, op) => sum + op.amount, 0);
  return account;
}

function ensureNoNegative(operations) {
  let total = 0;
  for (let i = operations.length - 1; i >= 0; i--) {
    const op = operations[i];
    if (total + op.amount < 0) {
      op.amount = -op.amount;
    }
    total += op.amount;
  }
}
