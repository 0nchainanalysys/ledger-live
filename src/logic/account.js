// @flow

// accounts business logic functions to work with libcore

import { BigNumber } from "bignumber.js";
import type { Account } from "@ledgerhq/live-common/lib/types";
import { getCryptoCurrencyById } from "@ledgerhq/live-common/lib/helpers/currencies";
import { decodeAccountId } from "@ledgerhq/live-common/lib/helpers/account";
import type { AccountData } from "@ledgerhq/live-common/lib/bridgestream/types";

export const importExistingAccount = ({
  id,
  currencyId,
  name,
  freshAddress,
  freshAddressPath,
  index,
  balance,
}: AccountData): Account => {
  const { xpub } = decodeAccountId(id);
  const currency = getCryptoCurrencyById(currencyId);
  const account: $Exact<Account> = {
    id,
    xpub,
    name,
    currency,
    index,
    // these fields will be completed as we will sync
    freshAddress,
    freshAddressPath,
    blockHeight: 0,
    balance: BigNumber(balance),
    operations: [],
    pendingOperations: [],
    unit: currency.units[0],
    lastSyncDate: new Date(),
  };
  return account;
};
