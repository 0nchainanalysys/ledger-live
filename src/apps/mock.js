// @flow
import { of, throwError } from "rxjs";
import {
  ManagerAppDepInstallRequired,
  ManagerAppDepUninstallRequired
} from "@ledgerhq/errors";
import { getDirectDep, getDependencies } from "./logic";
import { findCryptoCurrencyById } from "../currencies";
import type { ListAppsResult, AppOp, Exec, InstalledItem } from "./types";
import type { ApplicationVersion, DeviceInfo } from "../types/manager";

export const deviceInfo155 = {
  version: "1.5.5",
  isBootloader: false,
  isOSU: false,
  managerAllowed: false,
  mcuVersion: "1.7",
  pinValidated: false,
  providerId: 1,
  majMin: "1.5",
  targetId: 823132164
};

export const prettyActionPlan = (ops: AppOp[]) =>
  ops.map(op => (op.type === "install" ? "+" : "-") + op.name).join(", ");

export const prettyInstalled = (items: InstalledItem[]) =>
  items
    .map(({ name, updated }) => name + (updated ? "" : " (outdated)"))
    .join(", ");

export const parseInstalled = (installedDesc: string): InstalledItem[] =>
  installedDesc
    .split(",")
    .filter(Boolean)
    .map(a => {
      const trimmed = a.trim();
      const m = /(.*)\(outdated\)/.exec(trimmed);
      if (m) {
        return { name: m[1].trim(), updated: false };
      }
      return { name: trimmed, updated: true };
    });

export function mockListAppsResult(
  appDesc: string,
  installedDesc: string,
  deviceInfo: DeviceInfo
): ListAppsResult {
  const apps = appDesc
    .split(",")
    .map(a => a.trim())
    .filter(Boolean)
    .map((name, i) => {
      const o: ApplicationVersion = {
        id: i,
        app: i,
        name,
        version: "0.0.0",
        description: null,
        display_name: name,
        icon: "",
        picture: 0,
        notes: null,
        perso: "",
        hash: "",
        firmware: "",
        firmware_key: "",
        delete: "",
        delete_key: "",
        device_versions: [],
        se_firmware_final_versions: [],
        providers: [],
        date_creation: "",
        date_last_modified: ""
      };
      const currency = findCryptoCurrencyById(name);
      if (currency) {
        o.currency = currency;
      }
      return o;
    });
  const appByName = {};
  apps.forEach(app => {
    appByName[app.name] = app;
  });

  const installed = parseInstalled(installedDesc);

  return {
    apps,
    appByName,
    deviceInfo,
    installed
  };
}

export const mockExecWithInstalledContext = (
  installedInitial: InstalledItem[]
): Exec => {
  let installed = installedInitial.slice(0);
  return (appOp: AppOp, targetId: string | number, app: ApplicationVersion) => {
    if (appOp.name !== app.name) {
      throw new Error("appOp.name must match app.name");
    }

    if (
      getDependencies(app.name).some(dep => installed.some(i => i.name === dep))
    ) {
      return throwError(new ManagerAppDepUninstallRequired(""));
    }

    if (appOp.type === "install") {
      const dep = getDirectDep(app.name);
      if (dep) {
        const depInstalled = installed.find(i => i.name === dep);
        if (!depInstalled || !depInstalled.updated) {
          return throwError(new ManagerAppDepInstallRequired(""));
        }
      }
    }

    switch (appOp.type) {
      case "install":
        if (!installed.some(i => i.name === appOp.name)) {
          installed = installed.concat({ name: appOp.name, updated: true });
        }
        break;

      case "uninstall":
        installed = installed.filter(a => a.name !== appOp.name);
        break;
    }

    return of({ progress: 0 }, { progress: 0.5 }, { progress: 1 });
  };
};
