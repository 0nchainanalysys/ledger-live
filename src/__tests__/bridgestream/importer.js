// @flow
import { getFiatUnit } from "@ledgerhq/currencies";
import { makeChunks } from "../../bridgestream/exporter";
import {
  parseChunksReducer,
  areChunksComplete,
  chunksToResult
} from "../../bridgestream/importer";
import { genAccount } from "../../mock/account";

test("import", () => {
  const accounts = Array(3)
    .fill(null)
    .map((_, i) => genAccount("export_" + i));
  const arg = {
    accounts,
    exporterName: "test",
    exporterVersion: "0.0.0"
  };
  const chunks = makeChunks(arg);

  let data = [];
  [1, 2, 0, 3].forEach((nb, i) => {
    expect(areChunksComplete(data)).toBe(false);
    data = parseChunksReducer(data, chunks[nb]);
    expect(data.length).toBe(i + 1);
    data = parseChunksReducer(data, chunks[nb]);
    expect(data.length).toBe(i + 1); // chunk already existed
  });
  expect(areChunksComplete(data)).toBe(true);
  const res = chunksToResult(data);
  // $FlowFixMe
  expect(res.accounts).toMatchObject(
    accounts.map(a => ["account", a.id, a.name, a.currency.coinType])
  );
  expect(res).toMatchSnapshot();
});
