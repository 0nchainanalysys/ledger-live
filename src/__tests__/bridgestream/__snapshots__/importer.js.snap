// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`import 1`] = `
Object {
  "accounts": Array [
    Array [
      "account",
      "mock_account_export_0",
      "pQNKgto",
      "peercoin",
    ],
    Array [
      "account",
      "mock_account_export_1",
      "2KTQvcmdW2zTneCnPImZvgYLGi5rIBgc",
      "peercoin",
    ],
    Array [
      "account",
      "mock_account_export_2",
      "23PIvq83MrbdhquMsST6yH4PX3dBptsh",
      "peercoin",
    ],
  ],
  "meta": Object {
    "chunksFormatVersion": 2,
    "exporterName": "test",
    "exporterVersion": "0.0.0",
  },
}
`;
